controllers/scoring.py

```diff
@@ -0,0 +1,240 @@
+from magicbot import state, StateMachine, tunable, feedback
+from wpimath.geometry import Pose2d, Rotation2d, Translation2d
+import wpilib
+
+from components.gripper import Gripper
+from components.intake import Intake
+from components.arm import Arm, Setpoints, Setpoint
+from controllers.movement import Movement
+from utilities.game import (
+    GamePiece,
+    field_flip_pose2d,
+    get_double_substation,
+    get_single_substation,
+    field_flip_rotation2d,
+    BLUE_NODES,
+    Rows,
+)
+
+
+class ScoringController(StateMachine):
+    gripper: Gripper
+    intake: Intake
+    arm: Arm
+    movement: Movement
+
+    # how long before reaching pickup shelf to start closing claw
+    GRAB_PRE_TIME = tunable(0.5)
+    SCORE_PRE_TIME = tunable(0.1)
+    # amont of time before arriving at goal to stow arm before which
+    AUTO_STOW_CUTOFF = tunable(5)
+    # how long before arriving at the correct position for each action
+    # to move the arm to correct position or start intaking
+    PICKUP_CUBE_PREPARE_TIME = tunable(3)
+    SCORE_PREPARE_TIME = tunable(3)
+    PICKUP_CONE_PREPARE_TIME = tunable(3)
+
+    DESIRED_ROW = tunable(0)
+    DESIRED_COLUMN = tunable(0)
+
+    # swap the the side of the substation to test on a half field
+    swap_substation = tunable(False)
+
+    def __init__(self) -> None:
+        self.is_holding = GamePiece.NONE
+        self.autodrive = False
+        self.wants_piece = GamePiece.CONE
+        self.wants_to_intake = False
+        # use double substation shelf on right side from drivers pov
+        self.cone_pickup_side_right = False
+
+    def get_correct_autodrive_state(self) -> str:
+        cur_piece = self.get_current_piece()
+        if cur_piece is GamePiece.CONE or cur_piece is GamePiece.CUBE:
+            return "score"
+        elif self.wants_piece is GamePiece.CONE:
+            return "auto_pickup_cone"
+        elif self.wants_piece is GamePiece.CUBE:
+            return "auto_pickup_cube"
+        return "idle"
+
+    def goto_autodrive_state(self):
+        self.next_state(self.get_correct_autodrive_state())
+
+    @state(first=True)
+    def idle(self):
+        if self.autodrive:
+            self.goto_autodrive_state()
+        if self.wants_to_intake:
+            self.next_state("intaking")
+
+    @state
+    def intaking(self):
+        if self.autodrive:
+            self.goto_autodrive_state()
+        if not self.wants_to_intake:
+            self.next_state("idle")
```

Comment:
If this receives a request to autodrive and stop intaking, this will first go to the idle state, and then finally run the entirety of this state, potentially overwriting the next state with the grab_from_well state. Is this intended?

Reply:
yes, I think that behavior is fine